protos/tunneloffload.proto

```diff
@@ -0,0 +1,352 @@
+//
+enum ADD_TUNNEL_STATUS {
+  _SESSION_ACCEPTED = 0;
+  _SESSION_REJECTED = 1;
+  _SESSION_TABLE_FULL = 2;
+  _SESSION_TABLE_UNAVAILABLE = 3;
+  _SESSION_ALREADY_EXISTS= 4;
+}
+
+service ipTunnelService {
+  // Get which capabilities are available while using the  
+  rpc Capabilities(CapabilityRequest) returns (CapabilityResponse);
+
+  // Creation of IP Tunnel
+  // This API shuold be generic and allow creations of many IP tunnels
+  rpc createIpTunnel(stream ipTunnelRule) returns (ipTunnelResponses) {}
+
+  // Updating the tunnel,
+  // Only the tunnelId should be constant, all other parameters can be changed in the tunnel
+  rpc updateIpTunnel(stream ipTunnelRule) returns (ipTunnelResponses) {}
+
+  // Getting a tunnel by it's ID 
+  rpc getIpTunnel(tunnelId) returns (ipTunnelResponse) {} 
+
+  // Deleting a session by it's ID
+  rpc deleteIpTunnel(tunnelId) returns (ipTunnelResponse) {} 
+
+  // Getting all the ipTunnels currently configured
+  rpc getAllIpTunnels(tunnelRequestArgs) returns (ipTunnelResponses) {} // Contains the page size / etc
+} 
+
+// Capabilty request is empty since no paramteres are supplied to it,
+// all capabilities will be provided at response
+message CapabilityRequest {
+}
+
+
+enum ENC_TYPE {
+  _AES = 0;
+}
+
+enum AUTH_TYPE {
+  _SHA256 = 0;
+}
+
+message CapabilityResponse {
+  // We'll have capability for matching, and for every tunnel
+
+  message MatchCapabilities {
+    bool ingressInterfaceMatching = 1; // Is interface can be matched for encapsulation / decapsulation
+    bool vxlanMatching = 2; // Match with VXLAN VNI
+    bool geneveMatching = 3; // Match with geneve can happen
+    bool vrfMatching = 4; // Is vrf matching possible
+    bool tunnelMatching = 5; // Matching on tunnel
+    bool ipsecMatching = 6; // Can match on IPSec
+  }
+
+  MatchCapabilities matchCapabilities = 1;
+
+  message TunnelCapabilties {
+    bool ipsecOffload = 1; // Can the device offload ipsec
+    bool geneveOffload = 2; // Can the device offload geneve
+  }
+  
+  TunnelCapabilties tunnelCapabilities = 2;
+
+  message IPSecCapabilities {
+    repeated IPSecTunnelType tunnelTypeSupported = 1;
+    repeated AUTH_TYPE authSupported = 2;
+    repeated ENC_TYPE encryptionSupported = 3;
+  }
+  
+  IPSecCapabilities ipsecCapabilities = 3;
+  
+}
+
+enum ACTION {
+  // In this action packet will be forwarded right away
+  FORWARD = 0;
+  // In this action packet will continue processing, to other tunnel
+  RECIRCULATE = 1;
+}
+
+
+message MatchCriteria {
+    // In case it's not present, untagged traffic will be matched
+    string ingressInterface = 1; // Optional field, in which interface this tunnel will be encapsulated
+
+    MacPair macMatch = 2; // MAC of the packet itself
+
+    oneof ip_match {
+      IPV4Match ipv4Match = 3;
+      IPV6Match ipv6Match = 4;
+    }
+
+    string vrf = 5; // Optional field, not indicating VRF means that 
```

Comment:
Dont understand how the packet processing can match on the VRF - Is this a derived attribute based on the incoming logical interface that the packet arrived on ?